### ISSUE TITLE
Amounts not parsing correctly

### ISSUE THREAD

**Issue by 2426021684** [nlp-compromise/nlp_compromise#155]
`nlp_compromise.value("5 and a half").number` is equal to 6. It should be equal to 5.5. I am using the latest from head.

**Comment by 2426021684** [comment]
`nlp_compromise.value("5 1/2 grams").number` gives `null`

**Comment by spencermountain** [comment]
yeah, it'd be sweet to handle `fractions -> decimals`. that's a really neat idea.

**Comment by spencermountain** [comment]
(still not working!)
https://runkit.com/583f2a3dc534df0014136221/58d5d37154c1be00146c4b0f

**Comment by LakshSingla** [comment]
@spencermountain I would like to contribute towards this bugfix (in case this issue is still open) as my first contribution to this repository. Can you please give a brief overview of what needs to be done and how to accomplish it :smiley:

**Comment by spencermountain** [comment]
hey @LakshSingla that's lovely, thank you.
yeah, there's a few things that need to happen. 

1) 'and a half' (and whatever variants) should be tagged as a `Value`, so that they appear in the results of `doc.values()` - i think this may be working already, but worth playing around with. I'm thinking about `half a million` vs `four million and a half`, and so on...

2) The harder work is in the [parseText method](https://github.com/spencermountain/compromise/blob/master/src/subset/values/parseText/index.js#L28) and sorry, this is a bit of a hairy method. Feel free to refactor it, if you want to clean it up.
You can see it parses `two hundred and fifty million` into essentialy:
```js
sum+={ ones: 2 } * multiplier
sum+={ tens: 50, ones: 6 } * multiplier
```
the 'x and a half' should probably add `0.5` somewhere in there, although it may be complicated.
maybe like this? I don't know.
```js
sum+=({ ones: 2 } + 0.5 )* multiplier
sum+=({ tens: 50, ones: 6 } + 0.5 ) * multiplier
```
pseudocode, but i hope that's get's something across. That's how i'd go about doing it.
lemme know if i can help
cheers

**Comment by LakshSingla** [comment]
@spencermountain Regarding `and a half` issue here's what I suggest

| Syntax          | Semantics |
|:-----------------:|:------------------:|
| `\w+ and a? half\`          | Add half to the sum    |
| `\(some cardinal) and a? half (some multiple)\`| Add half to that cardinal and add to the sum after multiplying it with that multiple |

Are there are any more cases that we should consider? Or is there any modification with respect to these cases? Also, where should the test cases be added?

**Comment by spencermountain** [comment]
great, yeah, the other variant i can think of is `"half a (multiple)" `→ `multiple/2`?

it's cool if you wanna just plow away at it, I'll merge whatever. 
related tests are [here](https://github.com/spencermountain/compromise/blob/master/test/unit/subset/value/toNumber.test.js)
thanks!

**Comment by LakshSingla** [comment]
Sure thing, is there a way to document it for general reference later on?

**Comment by LakshSingla** [comment]
@spencermountain Till the time we are working on this issue, lets use my above comment as a place for documenting regarding this issue, as soon as I will make some substantial progress, I will update the tests and the docs. I have introduced `a? half` as a standalone identifier. Since I am new to this repo, will converting it into a value cause issue with other parts of the sentence like `half loaf` etc. ?

**Comment by spencermountain** [comment]
cool, okay.
i would do this, to guard against that usage, using the compromise match syntax:
```
doc.match('half a #Value').tag('#Value')
doc.match('#Value and a half').tag('#Value')
```
you can put that anywhere in the tagger. I think i have some of it already [here](https://github.com/spencermountain/compromise/blob/master/src/tagger/steps/23-corrections.js#L73)

deeper into that parseText method, it's a string, so you'll need to do a regex for it

**Comment by LakshSingla** [comment]
Alright, thanks a lot. I have written the logic for `^a? half$` and `\^w+ and a half$` (updated from `a? half` and `w+ and a? half$`, but have yet to test them. I figured a test driven development approach (TDD) would better in this case. Hence I will write the tests first. While I work on other functionality, it would be helpful if you review those cases and tell if they are inadequate or not. 
PS: Please review the changes made (mentioned in the first line) and I will update my earlier comment accordingly.

**Comment by spencermountain** [comment]
hey @LakshSingla yeah sounds like you're doing the right thing. The parse method should receive a trimmed string, like you suspected.
If something is up, or there are some terms missing, you can look at [.values()](https://github.com/spencermountain/compromise/blob/master/src/subset/values/index.js#L124) to make sure the correct terms are being grabbed.
Other than that, I trust your judgement about the proper behaviours, and tests.

That parse method is pretty weird, I admit. I remember it being hard to reason about

**Comment by LakshSingla** [comment]
Alright, thanks for your help. I will push the commit with reference to `a half` test cases today and mention it here in the comment. The current code for `a half` and related does not work and I suspect it may have something to do with `values()` you mentioned above. Also regarding `#1` it is working on values such as `seven and a half` or `seven eleven and a half` but failing on tests like `seven and a half eleven`. I will look into that today too.

**Comment by LakshSingla** [comment]
Which subset of strings does [.values()](https://github.com/spencermountain/compromise/blob/master/src/subset/values/index.js#L124) recieve , i.e. what is `r`? I am presuming the full string. I don't see any splitting based on `half` containing words. Is it already implemented or should I implement it here. Why do some of the preliminary tests like `['seven eleven and a half', '7 11.5']`pass then :confused: ( Is it our luck based on previous splitting patterns? )

**Comment by spencermountain** [comment]
yeah, `r` is a `Document` object. (I think it used to be called **Result**)
it's the same thing users will work with. You can see there's `splitOn` etc, methods [here](https://beta.observablehq.com/@spencermountain/compromise-api#Terms). Any of those should work.

**Comment by LakshSingla** [comment]
@spencermountain Sorry for late response :disappointed: I am creating a PR for my progress till now, which isn't that much. I am only able to cover `#1` case. For standalone `half` or `a half`, the code is failing. It would be helpful if you can take a look over and tell where the problem might lie. Also, against which branch should I make a PR?

**Comment by spencermountain** [comment]
cool, yeah a PR to master is fine,
don't worry about the failing edge-cases. That's fine, just comment-out that test.

**Comment by ampyard-akshay-sushir** [comment]
can any one works on this issue? I want to take a look at it. How can I contribute?

**Comment by spencermountain** [comment]
thank you @ampyard-akshay-sushir - i think a lot of these examples have been solved by https://github.com/spencermountain/compromise/pull/799 
note that the api has changed a lot in the 5 years since this ticket was created.
maybe it would be helpful to find any issues you can with natural-language fractions, and create a new ticket, then we can discuss how to fix them, sound good?
cheers!
